design/capm3-remediation-controller-proposal.md

```diff
@@ -0,0 +1,382 @@
+for my node when I find it to be unhealthy.
+
+#### Story 3
+
+As an admin, I would like to view historical data about the health of my
+cluster nodes for debugging purposes.
+
+#### Story 4
+
+As an admin, I should be able to mark ready nodes to be unusable.
+
+### Design Details
+
+CAPI machine health check allows the infrastructure provider to implement its
+own remediation controller to reconcile External Machine Remediation (EMR) CRs.
+EMRs are generated using infrastructure provider defined External Remediation
+Template.
+
+CAPM3 needs a new controller, the Remediation Controller to watch and reconcile
+Metal3Remediation objects. New controller does not have effect on existing APIs
+and adding of new remediation features does not include changes in other CAPM3
+controllers. This design adds two new CRDs  Metal3Remediation and
+Metal3RemediationTemplate.
+
+#### Remediation Controller
+
+* MRC watches for the presence of Metal3Remediation CR.
+* Based on the remediation strategy defined in ```.spec.strategy``` in
+  Metal3Remediation, RC uses BMO apis to get hosts back into a healthy or
+  manageable state.
+* RC uses ```.status.phase``` to save the states of the remediation. Available
+  states are running, rebooting, power off, failed.
+* After RC have finished its remediation, it will wait for the Metal3Remediation
+  CR to be removed. (When using CAPI MachineHealthCheck controller,
+  MHC will noticed the Node becomes healthy and deletes the instantiated
+  MachineRemediation CR.).
+* If RCs timeout for Node to become healthy expires, it annotates BareMetalHost
+  with capi.metal3.io/unhealthy annotation, sets BMH Online field to False and
+  removes the Machine object.
+
+#### Remediation operations
+
+* BMO reboot annotation: reboot.metal3.io
+* BMO power off/on annotation: reboot.metal3.io/remediation-controller-power-off
+  Note: These APIs are not responsible for managing unprovisioned hosts.
+* CAPM3 unhealthy annotation: capi.metal3.io/unhealthy
+  Note: Prevent annotated BMH to be selected as a host for a Machine.
+
+#### Watch
+
+* Watch Metal3Remediation CR resources.
+* Watch BareMetalHost CR resources.
+
+#### Reconcile
+
+* Fetch the Machine matching the name of the Metal3Remediation and operate over
+  BareMetalHost objects.
+
+#### Metal3Remediation CRD
+
+```yaml
+apiVersion: remediation.metal3.io/v1alpha4
+kind: Metal3Remediation
+metadata:
+  name: NAME_OF_UNHEALTHY_MACHINE
+  namespace: NAMESPACE_OF_UNHEALTHY_MACHINE
+   ownerReferences:
+apiVersion:cluster.x-k8s.io/v1alphaX
+kind: MachineHealthCheck
+name: REMEDIATION_GROUP
+uid: ...
+  spec:
+    strategy:
+      type: reboot
+      rebootLimit: 3
+      timeout: 300s
+  status:
+    phase: rebooting
+    retryCount: 1
+```
+
+#### Metal3RemediationTemplate CRD
+
+Metal3RemediationTemplate is used in the clusters that are using CAPI
+MachineHealthCheck CR for detecting unhealthy Machines. MHC instantiates the
+template generating Metal3Remediation.
```

Comment:
How does the MHC know which template to use?

Reply:
This will be a field in the MachineHealthCheck CR that an admin should create

Reply:
These are the fields in MHC.
```
remediationTemplate:
        kind: Metal3RemediationTemplate
        apiVersion: remediation.metal3.io/v1alphaX
        name: exampleRemediation
```

Reply:
Is instantiating a template like this already a feature of the MHC, or is implementing that part of the work being described?

Reply:
There is proposal [3190](https://github.com/kubernetes-sigs/cluster-api/pull/3190) for adding this feature.

Reply:
OK, we need to somehow clearly indicate that having that work accepted is a prerequisite of this work. Unless I misunderstand and it is not a prerequisite?

Reply:
Yes, I think dependencies section was used here to indicate that. But I also agree that we should explicitly mention that it is a prerequisite.

Reply:
Ah, sorry, I missed that dependency on my earlier review.